lib/runtime/life/resolver.go

```diff
@@ -513,6 +573,757 @@ func ext_offchain_index_set_version_1(vm *exec.VirtualMachine) int64 {
+		return 0
+	}
+
+	var kp crypto.Keypair
+
+	if seed.Exists() {
+		kp, err = ed25519.NewKeypairFromMnenomic(string(seed.Value()), "")
+	} else {
+		kp, err = ed25519.GenerateKeypair()
+	}
+
+	if err != nil {
+		logger.Warn("[ext_crypto_ed25519_generate_version_1] cannot generate key", "error", err)
+		return 0
+	}
+
+	ks, err := ctx.Keystore.GetKeystore(id)
+	if err != nil {
+		logger.Warn("[ext_crypto_ed25519_generate_version_1]", "name", id, "error", err)
+		return 0
+	}
+
+	ks.Insert(kp)
+
+	ret, err := toWasmMemorySized(memory, kp.Public().Encode(), 32)
+	if err != nil {
+		logger.Warn("[ext_crypto_ed25519_generate_version_1] failed to allocate memory", "error", err)
+		return 0
+	}
+
+	logger.Debug("[ext_crypto_ed25519_generate_version_1] generated ed25519 keypair", "public", kp.Public().Hex())
+	return int64(ret)
+}
+
+func ext_crypto_ed25519_sign_version_1(vm *exec.VirtualMachine) int64 {
+	logger.Trace("[ext_crypto_ed25519_sign_version_1] executing...")
+
+	keyTypeID := vm.GetCurrentFrame().Locals[0]
+	key := vm.GetCurrentFrame().Locals[1]
+	msg := vm.GetCurrentFrame().Locals[2]
+	memory := vm.Memory
+
+	id := memory[keyTypeID : keyTypeID+4]
+
+	pubKeyData := memory[key : key+32]
+	pubKey, err := ed25519.NewPublicKey(pubKeyData)
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_sign_version_1] failed to get public keys", "error", err)
+		return 0
+	}
+
+	ks, err := ctx.Keystore.GetKeystore(id)
+	if err != nil {
+		logger.Warn("[ext_crypto_ed25519_sign_version_1]", "name", id, "error", err)
+		ret, _ := toWasmMemoryOptional(memory, nil)
+		return ret
+	}
+
+	var ret int64
+	signingKey := ks.GetKeypair(pubKey)
+	if signingKey == nil {
+		logger.Error("[ext_crypto_ed25519_sign_version_1] could not find public key in keystore", "error", pubKey)
+		ret, err = toWasmMemoryOptional(memory, nil)
+		if err != nil {
+			logger.Error("[ext_crypto_ed25519_sign_version_1] failed to allocate memory", err)
+			return 0
+		}
+		return ret
+	}
+
+	sig, err := signingKey.Sign(asMemorySlice(memory, msg))
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_sign_version_1] could not sign message")
+	}
+
+	ret, err = toWasmMemoryFixedSizeOptional(memory, sig)
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_sign_version_1] failed to allocate memory", err)
+		return 0
+	}
+
+	return ret
+}
+
+func ext_crypto_ed25519_verify_version_1(vm *exec.VirtualMachine) int64 {
+	logger.Trace("[ext_crypto_ed25519_verify_version_1] executing...")
+
+	sig := vm.GetCurrentFrame().Locals[0]
+	msg := vm.GetCurrentFrame().Locals[1]
+	key := vm.GetCurrentFrame().Locals[2]
+	memory := vm.Memory
+	sigVerifier := ctx.SigVerifier
+
+	signature := memory[sig : sig+64]
+	message := asMemorySlice(memory, msg)
+	pubKeyData := memory[key : key+32]
+
+	pubKey, err := ed25519.NewPublicKey(pubKeyData)
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_verify_version_1] failed to create public key")
+		return 0
+	}
+
+	if sigVerifier.IsStarted() {
```

Comment:
What `sigVerifier.IsStarted()` means?

Reply:
I'm not sure, I used this based on code from wasmer.